src/cpu/pred/btb_entry.hh

```diff
@@ -0,0 +1,207 @@
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Declaration of a BTB entry and BTB indexing policy.
+ */
+
+#ifndef __CPU_PRED_BTB_ENTRY_HH__
+#define __CPU_PRED_BTB_ENTRY_HH__
+
+#include <vector>
+
+#include "arch/generic/pcstate.hh"
+#include "base/cache/cache_entry.hh"
+#include "base/intmath.hh"
+#include "base/types.hh"
+#include "cpu/static_inst.hh"
+#include "mem/cache/tags/indexing_policies/set_associative.hh"
+#include "params/BTBIndexingPolicy.hh"
+
+namespace gem5::branch_prediction
+{
+
+class BTBTagTypes
+{
+  public:
+    struct KeyType
+    {
+        Addr address;
+        ThreadID tid;
+    };
+    using Params = BaseIndexingPolicyParams;
+};
+
+class BTBEntry : public CacheEntry
+{
+  public:
+    using IndexingPolicy = BTBIndexingPolicy;
+    using KeyType = BTBTagTypes::KeyType;
+    using TagExtractor = std::function<Addr(Addr)>;
+
+    /** Default constructor */
+    BTBEntry(TagExtractor ext)
+        : CacheEntry(ext), inst(nullptr), tid(0)
+    {}
+
+    /** Update the entry in the BTB */
+    void
+    update(ThreadID _tid,
```

Comment:
Since it seems that we only update() right after calling insert, the _tid assignment is redundant and should probably be dropped.

Reply:
Removed tid argument.